gr-fec/lib/awgn_bp.cc

```diff
@@ -99,38 +104,70 @@ void awgn_bp::spa_initialize()
 
 void awgn_bp::update_chks()
 {
-    double product, _prdct;
+    double tanh_prod, atanh, x;
     int v;
+    int w;
+    //(step 1) of LLR-BP tanh algo.
+    // iterate over rows of check matrix
     for (int chk = 0; chk < M; chk++) {
-        product = double(1.0);
-        for (int i = 0; i < num_mlist[chk]; i++) {
-            v = mlist[chk][i] - 1;
-            product = product * double(2 / (1 + Q[chk][v]) - 1);
-        }
+        // iterate over nonzero cols of check matrix
         for (int i = 0; i < num_mlist[chk]; i++) {
+            tanh_prod = double(1.0);
             v = mlist[chk][i] - 1;
-            _prdct = product / double(2 / (1 + Q[chk][v]) - 1);
-            R[chk][v] = double((1 - _prdct) / (1 + _prdct));
+            // compute product of sign and and tanh LLR's
+            for (int iprime = 0; iprime < num_mlist[chk]; iprime++) {
+                // exclude iprime from the product set
+                if (iprime == i) {
+                    continue;
+                }
+                w = mlist[chk][iprime] - 1;
+                // remove divide by 2 for manual tanh
+                x = Q[chk][w] / 2.0;
+                // x = std::abs(Q[chk][w]);
+                // clamp tanh input (this is some BS, FIXME)
+                if (x > 18) {
+                    x = 18;
+                }
+                if (x < -18) {
+                    x = -18;
+                }
```

Comment:
```suggestion
                x = branchless_clip(x, 18.0);
```

We can use [`branchless_clip()`](https://github.com/gnuradio/gnuradio/blob/main/gnuradio-runtime/include/gnuradio/math.h#L77) here.